src/lem/circuit.rs

```diff
@@ -728,61 +783,197 @@ impl Func {
+                        and(&mut cs.namespace(|| "if_eq.and.2"), not_dummy, &not_eq)?;
+
+                    let saved_slot = &mut next_slot.clone();
+                    recurse(
+                        &mut cs.namespace(|| "if_eq.true"),
+                        eq_block,
+                        &not_dummy_and_eq,
+                        saved_slot,
+                        bound_allocations,
+                        preallocated_outputs,
+                        g,
+                    )?;
+                    let saved_slot = &mut next_slot.clone();
+                    recurse(
+                        &mut cs.namespace(|| "if_eq.false"),
+                        else_block,
+                        &not_dummy_and_not_eq,
+                        saved_slot,
+                        bound_allocations,
+                        preallocated_outputs,
+                        g,
+                    )
+                }
+                Ctrl::MatchTag(match_var, cases, def) => {
                     let allocated_match_tag = bound_allocations.get(match_var)?.tag().clone();
-                    let mut concrete_path_vec = Vec::new();
-                    for (tag, op) in cases {
+                    let mut selector = Vec::new();
+                    for (tag, block) in cases {
                         let allocated_has_match = alloc_equal_const(
                             &mut cs.namespace(|| format!("{tag}.alloc_equal_const")),
                             &allocated_match_tag,
                             tag.to_field::<F>(),
                         )
                         .with_context(|| "couldn't allocate equal const")?;
 
-                        let concrete_path_and_has_match = and(
+                        let not_dummy_and_has_match = and(
                             &mut cs.namespace(|| format!("{tag}.and")),
-                            concrete_path,
+                            not_dummy,
                             &allocated_has_match,
                         )
                         .with_context(|| "failed to constrain `and`")?;
 
-                        concrete_path_vec.push(allocated_has_match);
+                        selector.push(allocated_has_match);
 
                         let saved_slot = &mut next_slot.clone();
                         recurse(
                             &mut cs.namespace(|| format!("{}", tag)),
-                            op,
-                            &concrete_path_and_has_match,
+                            block,
+                            &not_dummy_and_has_match,
                             saved_slot,
                             bound_allocations,
                             preallocated_outputs,
                             g,
                         )?;
                     }
 
+                    match def {
+                        Some(def) => {
+                            let default = selector.iter().all(|b| !b.get_value().unwrap());
```

Comment:
Probably not. This is okay because I'm just choosing a value for the witness based on previous witness values. I'm not building circuit from witness values. The idea is that you allocate a boolean unconstrained, only to constrain it later by the popcount (selector or whatever)